### PR TITLE
Allow disconnect from EventEmiter using object

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR adds options to disconnect all events using the object itself. After reading the code changed in #3449 it looks like disconnecting using objects is not working, but fails silently. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

close 33565

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

@tlambert03

### CODE REVIEW

**Comment by tlambert03** [comment]
thank you @Czaki!  I can confirm this fixes the test failures I was getting on disconnect in #3184 ... however, it looks like we've got some new test failures.  I have to say, I've always been a bit confused personally about the expectations in those `test_adding_removing_layer` tests that are now failing.  It kind of seems like we're asserting that if you add a callback to a layer, then add and remove it from a viewer, you should be left with _no_ callbacks?  @sofroniewn might have some insight there

**Comment by sofroniewn** [comment]
> I have to say, I've always been a bit confused personally about the expectations in those test_adding_removing_layer tests that are now failing. It kind of seems like we're asserting that if you add a callback to a layer, then add and remove it from a viewer, you should be left with no callbacks.

That is how the test is written right now as we at least previously had the case that there were no internal callbacks - see https://github.com/napari/napari/blob/cc06d317038bc11fa9e8d2d6528381901142517d/napari/_tests/test_adding_removing.py#L73-L76

I personally like having no internal callbacks, i find that might lead to a lot of complicated spagetti like code if care is not taken - HOWEVER - if we were to add some internal callbacks then we should probably adjust `test_adding_removing_layer` so that the expected number of callbacks is present after the layer gets created and then after adding and removing that same number of callbacks is still present (i.e. no more were added, no more were taken away).

We could split `test_adding_removing_layer` into two tests. Test 1 would assert that a new layer has the expected number of callbacks (right now zero). Test two would store the number of callbacks a layer has when newly created (right now zero) and then add and remove the layer and compare how many callbacks are there. This would separate out these two concepts and probably be cleaner. Thoughts @tlambert03? Does that clarify?

**Comment by Czaki** [comment]
This case is just fixed. I add missed direct test for EventEmmiter.

**Comment by tlambert03** [comment]
Thanks @sofroniewn 

> I personally like having no internal callbacks, 

oh I 100% agree here. we shouldn't change that.  rarely a need for `self.events.connect(self.whatever)`
looking at the tests again, I'm actually not seeing what confused me before!  (there was a moment I thought that external callbacks were being nuked... but I see that's also explicitly tested.)

anyway, don't mind me! 😅

**Comment by Czaki** [comment]
Again `test_roll_transpose_update[Image-data1-2]` failed in async mode. I do not have idea why and how it could interact with this PR.

**Comment by sofroniewn** [comment]
> Again test_roll_transpose_update[Image-data1-2] failed in async mode. I do not have idea why and how it could interact with this PR.

I think we're going to add the skip sync anywhere there see convo in https://github.com/napari/napari/pull/3519#issuecomment-955787898

can you add here? then maybe we can get all tests passing and merge. cc @jaimergp @perlman too (if this merges first then i can just merge #3519 without edit)

**Comment by tlambert03** [comment]
i do have one quick question here @Czaki.  Is this essentially a new feature to support that line that was added in #3217?  Or do we actually use that pattern `emitter.disconnect(object_not_callback)` anywhere else?

**Comment by Czaki** [comment]
In this PR I use it to simplify `disconnect_events` (which is not tested). Maybe it could allow simplifying other parts of code.  

But maybe a better solution is to remove the mentioned line and modify the test added in this PR.

File: napari/_tests/test_viewer.py
Comment (sofroniewn) [inline napari/_tests/test_viewer.py]:
```suggestion
```

**Review by sofroniewn — APPROVED** [review]
LGTM! Thanks!! Will merge tomorrow unless I hear otherwise

**Comment by sofroniewn** [comment]
> In this PR I use it to simplify disconnect_events (which is not tested). Maybe it could allow simplifying other parts of code.
>
> But maybe a better solution is to remove the mentioned line and modify the test added in this PR.

@tlambert03 do you have followup thoughts here? Just want to make sure this is the right approach before we merge

**Pull request merged**
